### PR TITLE
event: add the real value of damage taken

### PR DESCRIPTION
**Pull request by Julien2313**
(#244)
I haven't found any test example for this case (I found in parser_test.go something I could work on, but it wouldn't have tested my changes)

### CODE REVIEW

File: pkg/demoinfocs/game_events.go [hunk 1]
```diff
@@ -340,20 +340,39 @@ func (geh gameEventHandler) playerDeath(data map[string]*msg.CSVCMsg_GameEventKe
...
 func (geh gameEventHandler) playerHurt(data map[string]*msg.CSVCMsg_GameEventKeyT) {
+	userID := data["userid"].GetValShort()
+	player := geh.playerByUserID32(userID)
 	attacker := geh.playerByUserID32(data["attacker"].GetValShort())
+
 	wepType := common.MapEquipment(data["weapon"].GetValString())
-	userID := data["userid"].GetValShort()
 	wepType = geh.attackerWeaponType(wepType, userID)
 
+	health := int(data["health"].GetValByte())
```
Comment (markus-wa) [inline pkg/demoinfocs/game_events.go:350]: this should be `GetValShort`

File: pkg/demoinfocs/events/events.go [hunk 2]
```diff
@@ -367,7 +367,11 @@ type PlayerHurt struct {
 	WeaponString string            // Wrong for CZ, M4A1-S etc.
 	HealthDamage int
 	ArmorDamage  int
-	HitGroup     HitGroup
+	// damage taken are used to know the real value the player lost
+	// in example, if the player has been shot by an AWP with a HS and had 50 health left, this value would be 50, not ~420
+	HealthDamageTaken *int
+	ArmorDamageTaken  *int
```
Comment (markus-wa) [inline pkg/demoinfocs/events/events.go:373]: these shouldn't have to be pointers

File: pkg/demoinfocs/game_events.go [hunk 3]
```diff
@@ -340,20 +340,39 @@ func (geh gameEventHandler) playerDeath(data map[string]*msg.CSVCMsg_GameEventKe
...
-		HealthDamage: int(data["dmg_health"].GetValShort()),
-		ArmorDamage:  int(data["dmg_armor"].GetValByte()),
-		HitGroup:     events.HitGroup(data["hitgroup"].GetValByte()),
-		Weapon:       geh.getEquipmentInstance(attacker, wepType),
+		Player:            player,
+		Attacker:          attacker,
+		Health:            health,
+		Armor:             armor,
+		HealthDamage:      int(data["dmg_health"].GetValShort()),
+		ArmorDamage:       int(data["dmg_armor"].GetValByte()),
```
Comment (markus-wa) [inline pkg/demoinfocs/game_events.go:371]: let's reuse the vars from above here

```suggestion
		HealthDamage:      healthDamage,
		ArmorDamage:       armorDamage,
```

File: pkg/demoinfocs/game_events.go (hunk 1 above)
Comment (markus-wa) [reply pkg/demoinfocs/game_events.go:350]: sorry, I misunderstood what's going on here

File: pkg/demoinfocs/game_events.go (hunk 3 above)
Comment (markus-wa) [reply pkg/demoinfocs/game_events.go:371]: ignore me

File: pkg/demoinfocs/events/events.go (hunk 2 above)
Comment (Julien2313) [reply pkg/demoinfocs/events/events.go:373]: true

**Comment by markus-wa** [comment]
Thanks for the PR @Julien2313 ! - looks good overall, think we should just clean up some docs and remove the pointers :slightly_smiling_face:

**Comment by Julien2313** [comment]
> 
> 
> Thanks for the PR @Julien2313 ! - looks good overall, think we should just clean up some docs and remove the pointers 🙂

I wrote too much doc ?! xD

**Comment by markus-wa** [comment]
@Julien2313 not too much doc no, just about how it will be displayed in the generated doc. see https://github.com/markus-wa/demoinfocs-golang/pull/247#discussion_r519192771 (I did a ninja edit on that comment, so you might not have seen the docs bit about it)

**Comment by markus-wa** [comment]
actually, I think it's not even about godoc, it's about IDE auto-complete. pkg.go.dev might have handled it ok

**Comment by Julien2313** [comment]
> @Julien2313 not too much doc no, just about how it will be displayed in the generated doc. see [#247 (comment)](https://github.com/markus-wa/demoinfocs-golang/pull/247#discussion_r519192771) (I did a ninja edit on that comment, so you might not have seen the docs bit in there)

Hooo ok ok, got it

**Comment by Julien2313** [comment]
> excluding over-damage 
So well said ! ha ha

File: pkg/demoinfocs/events/events.go [hunk 4]
```diff
@@ -359,15 +359,17 @@ const (
...
+	Player            *common.Player // May be nil if the demo is partially corrupt (player is 'unconnected', see #156 and #172).
+	Attacker          *common.Player // May be nil if the player is taking world damage (e.g. fall damage) or if the demo is partially corrupt (player is 'unconnected', see #156 and #172).
+	Health            int
+	Armor             int
+	Weapon            *common.Equipment // May be EqUnknown for world-damage (falling / bomb).
+	WeaponString      string            // Wrong for CZ, M4A1-S etc.
+	HealthDamage      int
+	ArmorDamage       int
+	HealthDamageTaken int // HealthDamage excluding over-damage (e.g. if player has 5 health and is hit for 15 damage this would be 5 instead of 15)
+	ArmorDamageTaken  int // ArmorDamage excluding over-damage (e.g. if player has 5 armor and is hit for 15 armor damage this would be 5 instead of 15)
```
Comment (markus-wa) [inline pkg/demoinfocs/events/events.go:371]: nice one for catching the typo in my suggestion :clap: :smile:

File: pkg/demoinfocs/game_events.go [hunk 5]
```diff
@@ -340,20 +340,39 @@ func (geh gameEventHandler) playerDeath(data map[string]*msg.CSVCMsg_GameEventKe
...
+	player := geh.playerByUserID32(userID)
 	attacker := geh.playerByUserID32(data["attacker"].GetValShort())
+
 	wepType := common.MapEquipment(data["weapon"].GetValString())
-	userID := data["userid"].GetValShort()
 	wepType = geh.attackerWeaponType(wepType, userID)
 
+	health := int(data["health"].GetValByte())
+	armor := int(data["armor"].GetValByte())
+	healthDamageTaken := health
```
Comment (markus-wa) [inline pkg/demoinfocs/game_events.go:352]: hmm, I think this is wrong. shouldn't it be `int(data["dmg_health"].GetValShort())` by default and if `data["health"].GetValByte() == 0` then it should be `player.Health()`.

but it should never be `data["health"].GetValByte()` - right?

**Review by markus-wa — COMMENTED** [review]
hmm. I think there's an issue with the default value of `HealthDamageTaken` and `ArmorDamageTaken`

File: pkg/demoinfocs/game_events.go (hunk 5 above)
Comment (Julien2313) [reply pkg/demoinfocs/game_events.go:352]: Yes, you're right ! My bad

**Comment by Julien2313** [comment]
I added a check if the value is > 100, let me know if it's ok for you or not

**Comment by Julien2313** [comment]
Do you have somewhere an helper package, so I can add a max function max int ? To directly make the check at the assignation

File: pkg/demoinfocs/game_events.go [hunk 6]
```diff
@@ -340,20 +340,50 @@ func (geh gameEventHandler) playerDeath(data map[string]*msg.CSVCMsg_GameEventKe
...
+		}
+	} else {
+		if healthDamageTaken > 100 {
+			healthDamageTaken = 100
+		}
+
+		if armorDamageTaken > 100 {
+			armorDamageTaken = 100
+		}
+	}
```
Comment (markus-wa) [inline pkg/demoinfocs/game_events.go:374]: we should check this even if `player != nil`

Comment (markus-wa) [reply pkg/demoinfocs/game_events.go:374]: probably best to just move it in front of that if statement

**Comment by markus-wa** [comment]
> Do you have somewhere a helper package, so I can add a max function max int ? To directly make the check at the assignation.
(it will slow down the parser a little little bit as most of the time, player won't be nil, or I could let it like that)

I think it's fine as it is. it's a shame Go doesn't provide min/max functions for all data types in the stdlib, but I'd rather not add a new package or even function just for this.

**Review by markus-wa — APPROVED** [review]
LGTM :slightly_smiling_face: 

appreciate your efforts with all this, thanks a lot :pray:

**Comment by Julien2313** [comment]
> 
> 
> LGTM 🙂
> 
> appreciate your efforts with all this, thanks a lot 🙏

You are welcome, that's a wonderful library ;)

**Pull request merged**